demos/aks-nat-agic/azuredeploy.json

```diff
@@ -0,0 +1,2254 @@
+      ],
+      "properties": {
+        "roleDefinitionId": "[variables('contributorRoleId')]",
+        "description": "Assign the cluster user-defined managed identity contributor role on the resource group.",
+        "principalId": "[reference(variables('aksClusterUserDefinedManagedIdentityName')).principalId]",
+        "principalType": "ServicePrincipal",
+        "scope": "[resourceGroup().id]"
+      }
+    },
+    {
+      "type": "Microsoft.KeyVault/vaults",
+      "apiVersion": "2019-09-01",
+      "name": "[parameters('keyVaultName')]",
+      "location": "[parameters('location')]",
+      "dependsOn": [
+        "[variables('applicationGatewayUserDefinedManagedIdentityId')]",
+        "[variables('aadPodIdentityUserDefinedManagedIdentityId')]"
+      ],
+      "properties": {
+        "accessPolicies": [
+          {
+            "tenantId": "[reference(variables('applicationGatewayUserDefinedManagedIdentityId')).tenantId]",
+            "objectId": "[reference(variables('applicationGatewayUserDefinedManagedIdentityId')).principalId]",
+            "permissions": {
+              "secrets": ["get", "list"],
+              "certificates": ["get"]
+            }
+          },
+          {
+            "tenantId": "[reference(variables('aadPodIdentityUserDefinedManagedIdentityId')).tenantId]",
+            "objectId": "[reference(variables('aadPodIdentityUserDefinedManagedIdentityId')).principalId]",
+            "permissions": {
+              "secrets": ["get", "list"],
+              "certificates": ["get"]
+            }
+          }
+        ],
+        "sku": {
+          "family": "A",
+          "name": "standard"
+        },
+        "tenantId": "[subscription().tenantId]",
+        "networkAcls": {
+          "bypass": "AzureServices",
+          "defaultAction": "[parameters('keyVaultNetworkRuleSetDefaultAction')]"
+        },
+        "enabledForDeployment": false,
+        "enabledForDiskEncryption": false,
+        "enabledForTemplateDeployment": false,
+        "enableSoftDelete": false
+      },
+      "resources": [
+        {
+          "name": "Microsoft.Insights/default",
+          "type": "providers/diagnosticSettings",
+          "apiVersion": "2017-05-01-preview",
+          "dependsOn": [
+            "[variables('keyVaultId')]",
+            "[variables('workspaceId')]"
+          ],
+          "properties": {
+            "workspaceId": "[variables('workspaceId')]",
+            "logs": [
+              {
+                "category": "AuditEvent",
+                "enabled": true
+              }
+            ],
+            "metrics": [
+              {
+                "category": "AllMetrics",
+                "enabled": true
+              }
+            ]
+          }
+        },
+        {
+          "type": "providers/roleAssignments",
+          "apiVersion": "2020-04-01-preview",
+          "name": "[concat('Microsoft.Authorization/', guid(concat(resourceGroup().id), variables('readerRoleId')))]",
```

Comment:
roleAssignment names need to be a giud seeded with principal, role and scope.
```
guid(variables('aadPodIdentityUserDefinedManagedIdentityId'), variables('readerRoleId'), [and whatever your scope is here (hard to read with all the nesting)])
```

Reply:
Fixed

Reply:
Check the other roleAssignments in the template as well, they all need to follow the same pattern precisely.

Reply:
I think I fixed them